pkg/kubelet/metrics/metrics.go

```diff
@@ -822,6 +845,20 @@ var (
 			StabilityLevel: metrics.ALPHA,
 		},
 	)
+
+	// ImagePullDuration is a Histogram that tracks the duration (in seconds) it takes for an image to be pulled,
+	// including the time spent in the waiting queue of image puller.
+	// The metric is broken down by bucketed image size.
+	ImagePullDuration = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Subsystem:      KubeletSubsystem,
+			Name:           ImagePullDurationKey,
+			Help:           "",
+			Buckets:        podStartupDurationBuckets,
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"image_size_in_bytes"},
```

Comment:
It can be helpful to note how the bucket is measured in the label name.  Is it less than (or equal) to this image size?  Greater than?

Reply:
I reduced the number of pull duration buckets to 18.

Reply:
> It can be helpful to note how the bucket is measured in the label name. Is it less than (or equal) to this image size? Greater than?

The labels are now like "40GB-60GB", "GT100GB", so should be cleared enough?

Reply:
I'd recommend that being part of the key, not the value.

I would also recommend against ranges, as it makes it hard to add additional buckets over time (e.g. 40GB-60GB becomes 40GB-50GB and 50GB-60GB when split, which is breaking for users).

The key also says "in_bytes", but the value is not a number of bytes.

To summarize, I would recommend: `le_image_size_in_bytes` with float values in bytes (including +Inf).

Reply:
There is a tradeoff though: he problem with `le_image_size_in_bytes` is that you basically force create all the buckets in between which causes this metric to be uniformly high cardinality.

Reply:
The two approaches should have the same cardinality, no?

"image_size_in_bytes" -> "le_image_size_in_bytes"
"0-10MB" -> "10MB"
"10MB-100MB -> "100MB"
etc.

Reply:
Thats the case for fixed-bucket histograms generally.  I would prefer not to use other patterns for bucketing... If this is an `internal` metric, i would be fine with it.  But if we plan to stabilize it, being able to add buckets without breaking seems important.

Reply:
I think @logicalhan is saying we can omit buckets that have no observations if the buckets are exclusive (each range only has observations for that range) vs cumulative (each bucket has observations for that range, and for all buckets below it).  I would be curious what querying looks like for each method (and if it works at all with our 2d histogram).

Reply:
> I think @logicalhan is saying we can omit buckets that have no observations if the buckets are exclusive (each range only has observations for that range) vs cumulative (each bucket has observations for that range, and for all buckets below it). I would be curious what querying looks like for each method (and if it works at all with our 2d histogram).

oh even with le_image_size_in_bytes, each size bucket will be exclusive, no? For example, for an image pull of 5MB, it will be only tagged with le_image_size_in_bytes:10MB, but not the higher labels.  Otherwise we will need to emit multiple metric points like:

```
metrics.ImagePullDuration.WithLabelValues("10MB")).Observe(imagePullDuration.Seconds())
metrics.ImagePullDuration.WithLabelValues("100MB")).Observe(imagePullDuration.Seconds())
metrics.ImagePullDuration.WithLabelValues("1GB")).Observe(imagePullDuration.Seconds())
metrics.ImagePullDuration.WithLabelValues("5GB")).Observe(imagePullDuration.Seconds())
...
```

Reply:
@dashpole it is possible to omit the "buckets" without observations since they are represented as label values here. The impact would be similar to when we change bucket values in an histogram, the existing queries would just yield empty results, so it just really depends on how the query are built, but it shouldn't cause errors and alerts shouldn't fire.

Reply:
if you actually meant the real buckets and not just get rid of the label and underlying buckets, we would have to use native histograms

Reply:
> The impact would be similar to when we change bucket values in an histogram, the existing queries would just yield empty results, so it just really depends on how the queries are built, but it shouldn't cause errors and alerts shouldn't fire.

If you add buckets to a fixed-bucket histogram, all existing bucket series are untouched, since the number of observations less than or equal to a given value does not change.

> Otherwise we will need to emit multiple metric points like ...

Thats correct. It doesn't actually emit extra metrics, though. We are essentially reimplementing the logic of a fixed bucket histogram.  It doesn't necessarily increase cardinality since image sizes are likely roughly normally distributed, but might have cases where it has more buckets (e.g. 1 image at 100GB+, and no other images).

Reply:
> Thats correct. It doesn't actually emit extra metrics, though. We are essentially reimplementing the logic of a fixed bucket histogram. It doesn't necessarily increase cardinality since image sizes are likely roughly normally distributed, but might have cases where it has more buckets (e.g. 1 image at 100GB+, and no other images).

@dashpole In this case how do we query things like "P90 latency of images of size between 1GB and 5GB"? In current implementation I can just add a filter "image_size_in_bytes:"1GB-5GB". If we emit metrics to all buckets larger than the image size, how do we filter out the delta?

Reply:
> @dashpole In this case how do we query things like "P90 latency of images of size between 1GB and 5GB"? In current implementation I can just add a filter "image_size_in_bytes:"1GB-5GB". If we emit metrics to all buckets larger than the image size, how do we filter by image size range?

It's not natively supported since you're basically embedding a distribution in another distribution. I disagree with @dashpole on this one, I think your approach (i.e. using ranges in labels) is probably fine.

Reply:
> P90 latency of images of size between 1GB and 5GB

Wouldn't it just be: `histogram_quantile(0.90, (rate(image_pull_duration_seconds{image_size="5GB"}[5m]) - rate(image_pull_duration_seconds{image_size="1GB"}[5m])))`

I.e. "1GB-5GB" is just the number in the "5GB" bucket minus the number in the "1GB" bucket.

If we do choose to use labels with ranges, that just means we can't change them in the future. If we are OK with that, then we can take that approach.

Reply:
> If we do choose to use labels with ranges, that just means we can't change them in the future. If we are OK with that, then we can take that approach.

I don't think we will update this bucket ranges in future, or at least won't update it often. I prefer having the range buckets mostly to save the hassle of emitting metrics multiple times like the following:

```
metrics.ImagePullDuration.WithLabelValues("10MB")).Observe(imagePullDuration.Seconds())
metrics.ImagePullDuration.WithLabelValues("100MB")).Observe(imagePullDuration.Seconds())
metrics.ImagePullDuration.WithLabelValues("1GB")).Observe(imagePullDuration.Seconds())
metrics.ImagePullDuration.WithLabelValues("5GB")).Observe(imagePullDuration.Seconds())
...
```


And it will be easier to query with ranges.

Reply:
Thats fine.  Lets try and settle on buckets before the metric moves to stable.

Reply:
thanks @dashpole . Can you give this PR an approval if we are good to proceed?